### DISCUSSION TITLE
Implicit pattern match constructor lookup

### DISCUSSION THREAD

**Discussion by JonathanStarup** [flix/flix#6841] [Ideas]
It seems that in Ocaml you can do this magic constructor lookup

```ocaml
module A = struct type mylist = MyCons of int | MyNil end
type mylist2 = MyCons of string

match A.MyCons 1 with
  | MyCons x -> x
  | MyNil -> 123
```

Note that this doesn't work (`mylist2 doesn't have MyNil`)

```ocaml
module A = struct type mylist = MyCons of int | MyNil end
type mylist2 = MyCons of string

let f x = match x with
  | MyCons x -> x
  | MyNil -> assert true
```

but this does work

```ocaml
module A = struct type mylist = MyCons of int
type mylist2 = MyCons of string

let f x = match x with
  | MyCons x -> x
```
where `f` has type `mylist2 -> string`

So it seems that they import the constructors IF the scrutinee has a known data type at the time of inference of the match.

**Comment by magnus-madsen** [comment]
What happens with pairs etc? What if you match on a pair of two types  that both define MyFoo?

**Comment by magnus-madsen** [comment]
I guess I am still confused about how the resolution works-- when you match on a program variable `x`.

**Comment by JonathanStarup** [comment]
> What happens with pairs etc? What if you match on a pair of two types that both define MyFoo?

id guess different parts of the tuple have different imports

But yes it seems pretty wild
